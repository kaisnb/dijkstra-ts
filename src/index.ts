import PriorityQueue from "priorityqueuejs";

export type NodeKey = string | number;

export interface Edge<T> {
  node: T;
  weight: number;
}

interface QueueEntry<T> {
  node: T;
  cost: number;
}

export interface IGraphAdapter<T> {
  getKey?: (node: T) => NodeKey;
  getEdges: (node: T) => Edge<T>[];
}

const getNodeKey = <T>(node: T, adapter: IGraphAdapter<T>): NodeKey => {
  if (typeof adapter.getKey === "function") return adapter.getKey(node);
  if (typeof node === "string" || typeof node === "number") return node;
  throw new Error("Adapter must implement method getKey");
};

export const dijkstra = <T>(
  adapter: IGraphAdapter<T>,
  startNode: T,
  finishNode?: T
) => {
  const getKey = (node: T) => getNodeKey(node, adapter);
  const parents: Record<NodeKey, T> = Object.create(null);
  const costs: Record<NodeKey, number> = Object.create(null);
  const explored: Record<NodeKey, boolean> = Object.create(null);
  const prioQueue = new PriorityQueue<QueueEntry<T>>((a, b) => b.cost - a.cost);
  prioQueue.enq({ node: startNode, cost: 0 });

  do {
    let node = prioQueue.deq().node;
    let nodeKey = getKey(node);
    let cost = costs[nodeKey] || 0;

    explored[nodeKey] = true;

    // Early return when the shortest path in our
    // graph is already the finishNode
    if (undefined !== finishNode && nodeKey === getKey(finishNode)) break;

    const edges = adapter.getEdges(node);
    for (let i = 0; i < edges.length; i++) {
      const childNode = edges[i].node;
      const childNodeKey = getKey(childNode);
      let alt = cost + edges[i].weight;

      if (undefined === costs[childNodeKey] || alt < costs[childNodeKey]) {
        costs[childNodeKey] = alt;
        parents[childNodeKey] = node;

        if (!explored[childNodeKey]) {
          prioQueue.enq({ node: childNode, cost: alt });
        }
      }
    }
  } while (!prioQueue.isEmpty());

  return {
    costs,
    parents,
  };
};

export const findShortestPath = <T>(
  adapter: IGraphAdapter<T>,
  startNode: T,
  finishNode: T
) => {
  const getKey = (node: T) => getNodeKey(node, adapter);
  const { costs, parents } = dijkstra(adapter, startNode, finishNode);

  let optimalPath = [finishNode];
  let parent = parents[getKey(finishNode)];
  while (parent) {
    optimalPath.push(parent);
    parent = parents[getKey(parent)];
  }
  optimalPath.reverse();

  const results = {
    distance: costs[getKey(finishNode)],
    path: optimalPath,
  };

  return results;
};
