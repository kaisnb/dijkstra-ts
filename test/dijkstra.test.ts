import { dijkstra, Edge, findShortestPath, IGraphAdapter } from "../src";
import * as oGraph from "./graphs/object-graph";
import * as pGraph from "./graphs/primitive-graph";

describe("dijkstra", () => {
  const graphAdapter = new pGraph.GraphAdapter(pGraph.graph);

  const graphAdapterZeroWeight = new pGraph.GraphAdapter({
    [pGraph.startNode]: { A: 0, B: 2 },
    A: { [pGraph.finishNode]: 0 },
    B: { [pGraph.finishNode]: 2 },
    [pGraph.finishNode]: {},
  });

  class BrokenObjectGraphAdapter implements IGraphAdapter<oGraph.MyNodeImpl> {
    getEdges(n: oGraph.MyNodeImpl): Edge<oGraph.MyNodeImpl>[] {
      return n.edges.map(({ node, cost }) => ({ node, weight: cost }));
    }
  }

  it("should calculate all paths in an acyclic graph without negative edges", () => {
    const result = dijkstra(graphAdapter, pGraph.startNode);

    expect(result).toMatchSnapshot();
  });

  it("should early return when finishNode is found", () => {
    graphAdapter.getEdges = jest.spyOn(graphAdapter, "getEdges") as any;

    dijkstra(graphAdapter, pGraph.startNode, pGraph.finishNode);

    expect(graphAdapter.getEdges).toHaveBeenCalledTimes(4);
  });

  it("should use getKey when implemented", () => {
    const adapter = new oGraph.GraphAdapter();
    adapter.getKey = jest.spyOn(adapter, "getKey") as any;

    const result = dijkstra(adapter, oGraph.startNode);

    expect(result).toMatchSnapshot();
    expect(adapter.getKey).toBeCalledTimes(17);
  });

  it("should throw when getKey is not implemented", () => {
    expect(() =>
      dijkstra(new BrokenObjectGraphAdapter(), oGraph.startNode)
    ).toThrow("Adapter must implement method getKey");
  });

  describe("findShortestPath", () => {
    it("should return distance 8", () => {
      const result = findShortestPath(
        graphAdapter,
        pGraph.startNode,
        pGraph.finishNode
      );

      expect(result).toMatchSnapshot();
      expect(result.distance).toBe(8);
    });

    it("should return distance 0", () => {
      const result = findShortestPath(
        graphAdapterZeroWeight,
        pGraph.startNode,
        pGraph.finishNode
      );

      expect(result).toMatchSnapshot();
      expect(result.distance).toBe(0);
    });
  });
});
