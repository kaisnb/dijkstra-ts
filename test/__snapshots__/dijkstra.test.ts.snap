// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dijkstra findShortestPath should return distance 0 1`] = `
Object {
  "distance": 0,
  "path": Array [
    "start",
    "A",
    "finish",
  ],
}
`;

exports[`dijkstra findShortestPath should return distance 8 1`] = `
Object {
  "distance": 8,
  "path": Array [
    "start",
    "A",
    "D",
    "finish",
  ],
}
`;

exports[`dijkstra should calculate all paths in an acyclic graph without negative edges 1`] = `
Object {
  "costs": Object {
    "A": 5,
    "B": 2,
    "C": 9,
    "D": 7,
    "finish": 8,
  },
  "parents": Object {
    "A": "start",
    "B": "start",
    "C": "A",
    "D": "A",
    "finish": "D",
  },
}
`;

exports[`dijkstra should use getKey when implemented 1`] = `
Object {
  "costs": Object {
    "A": 5,
    "B": 2,
    "C": 9,
    "D": 7,
    "finish": 8,
  },
  "parents": Object {
    "A": MyNodeImpl {
      "edges": Array [
        MyEdgeImpl {
          "cost": 5,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 4,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 6,
                      "node": MyNodeImpl {
                        "edges": Array [
                          MyEdgeImpl {
                            "cost": 1,
                            "node": MyNodeImpl {
                              "edges": Array [],
                              "name": "finish",
                            },
                          },
                        ],
                        "name": "D",
                      },
                    },
                    MyEdgeImpl {
                      "cost": 3,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "C",
                },
              },
              MyEdgeImpl {
                "cost": 2,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 1,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "D",
                },
              },
            ],
            "name": "A",
          },
        },
        MyEdgeImpl {
          "cost": 2,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 8,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 4,
                      "node": MyNodeImpl {
                        "edges": Array [
                          MyEdgeImpl {
                            "cost": 6,
                            "node": MyNodeImpl {
                              "edges": Array [
                                MyEdgeImpl {
                                  "cost": 1,
                                  "node": MyNodeImpl {
                                    "edges": Array [],
                                    "name": "finish",
                                  },
                                },
                              ],
                              "name": "D",
                            },
                          },
                          MyEdgeImpl {
                            "cost": 3,
                            "node": MyNodeImpl {
                              "edges": Array [],
                              "name": "finish",
                            },
                          },
                        ],
                        "name": "C",
                      },
                    },
                    MyEdgeImpl {
                      "cost": 2,
                      "node": MyNodeImpl {
                        "edges": Array [
                          MyEdgeImpl {
                            "cost": 1,
                            "node": MyNodeImpl {
                              "edges": Array [],
                              "name": "finish",
                            },
                          },
                        ],
                        "name": "D",
                      },
                    },
                  ],
                  "name": "A",
                },
              },
              MyEdgeImpl {
                "cost": 7,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 1,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "D",
                },
              },
            ],
            "name": "B",
          },
        },
      ],
      "name": "start",
    },
    "B": MyNodeImpl {
      "edges": Array [
        MyEdgeImpl {
          "cost": 5,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 4,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 6,
                      "node": MyNodeImpl {
                        "edges": Array [
                          MyEdgeImpl {
                            "cost": 1,
                            "node": MyNodeImpl {
                              "edges": Array [],
                              "name": "finish",
                            },
                          },
                        ],
                        "name": "D",
                      },
                    },
                    MyEdgeImpl {
                      "cost": 3,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "C",
                },
              },
              MyEdgeImpl {
                "cost": 2,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 1,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "D",
                },
              },
            ],
            "name": "A",
          },
        },
        MyEdgeImpl {
          "cost": 2,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 8,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 4,
                      "node": MyNodeImpl {
                        "edges": Array [
                          MyEdgeImpl {
                            "cost": 6,
                            "node": MyNodeImpl {
                              "edges": Array [
                                MyEdgeImpl {
                                  "cost": 1,
                                  "node": MyNodeImpl {
                                    "edges": Array [],
                                    "name": "finish",
                                  },
                                },
                              ],
                              "name": "D",
                            },
                          },
                          MyEdgeImpl {
                            "cost": 3,
                            "node": MyNodeImpl {
                              "edges": Array [],
                              "name": "finish",
                            },
                          },
                        ],
                        "name": "C",
                      },
                    },
                    MyEdgeImpl {
                      "cost": 2,
                      "node": MyNodeImpl {
                        "edges": Array [
                          MyEdgeImpl {
                            "cost": 1,
                            "node": MyNodeImpl {
                              "edges": Array [],
                              "name": "finish",
                            },
                          },
                        ],
                        "name": "D",
                      },
                    },
                  ],
                  "name": "A",
                },
              },
              MyEdgeImpl {
                "cost": 7,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 1,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "D",
                },
              },
            ],
            "name": "B",
          },
        },
      ],
      "name": "start",
    },
    "C": MyNodeImpl {
      "edges": Array [
        MyEdgeImpl {
          "cost": 4,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 6,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 1,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "D",
                },
              },
              MyEdgeImpl {
                "cost": 3,
                "node": MyNodeImpl {
                  "edges": Array [],
                  "name": "finish",
                },
              },
            ],
            "name": "C",
          },
        },
        MyEdgeImpl {
          "cost": 2,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 1,
                "node": MyNodeImpl {
                  "edges": Array [],
                  "name": "finish",
                },
              },
            ],
            "name": "D",
          },
        },
      ],
      "name": "A",
    },
    "D": MyNodeImpl {
      "edges": Array [
        MyEdgeImpl {
          "cost": 4,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 6,
                "node": MyNodeImpl {
                  "edges": Array [
                    MyEdgeImpl {
                      "cost": 1,
                      "node": MyNodeImpl {
                        "edges": Array [],
                        "name": "finish",
                      },
                    },
                  ],
                  "name": "D",
                },
              },
              MyEdgeImpl {
                "cost": 3,
                "node": MyNodeImpl {
                  "edges": Array [],
                  "name": "finish",
                },
              },
            ],
            "name": "C",
          },
        },
        MyEdgeImpl {
          "cost": 2,
          "node": MyNodeImpl {
            "edges": Array [
              MyEdgeImpl {
                "cost": 1,
                "node": MyNodeImpl {
                  "edges": Array [],
                  "name": "finish",
                },
              },
            ],
            "name": "D",
          },
        },
      ],
      "name": "A",
    },
    "finish": MyNodeImpl {
      "edges": Array [
        MyEdgeImpl {
          "cost": 1,
          "node": MyNodeImpl {
            "edges": Array [],
            "name": "finish",
          },
        },
      ],
      "name": "D",
    },
  },
}
`;
